client/network/src/block_request_handler.rs

```diff
@@ -127,16 +178,91 @@ impl <B: BlockT> BlockRequestHandler<B> {
 
 		let direction = Direction::from_i32(request.direction)
 			.ok_or(HandleRequestError::ParseDirection)?;
+
+		let key = SeenRequestsKey {
+			peer: *peer,
+			max_blocks,
+			direction,
+			from: from_block_id.clone(),
+		};
+
+		let mut reputation_changes = Vec::new();
+
+		if let Some(requests) = self.seen_requests.get_mut(&key) {
+			*requests = requests.saturating_add(1);
+
+			if *requests > MAX_NUMBER_OF_SAME_REQUESTS_PER_PEER {
+				reputation_changes.push(rep::SAME_REQUEST);
+			}
+		} else {
+			self.seen_requests.put(key, 1);
+		}
+
+		debug!(
+			target: LOG_TARGET,
+			"Handling block request from {}: Starting at `{:?}` with maximum blocks \
+			 of `{}` and direction `{:?}`.",
+			peer,
+			from_block_id,
+			max_blocks,
+			direction,
+		);
+
 		let attributes = BlockAttributes::from_be_u32(request.fields)?;
+
+		let result = if reputation_changes.is_empty() {
+			let block_response = self.get_block_response(
+				&mut reputation_changes,
+				attributes,
+				from_block_id,
+				direction,
+				max_blocks,
+			)?;
+
+			let mut data = Vec::with_capacity(block_response.encoded_len());
+			block_response.encode(&mut data)?;
+
+			Ok(data)
+		} else {
+			Err(())
+		};
+
+		pending_response.send(OutgoingResponse {
+			result,
+			reputation_changes,
+		}).map_err(|_| HandleRequestError::SendResponse)
+	}
+
+	fn get_block_response(
+		&self,
+		reputation_changes: &mut Vec<ReputationChange>,
+		attributes: BlockAttributes,
+		mut block_id: BlockId<B>,
+		direction: Direction,
+		max_blocks: usize,
+	) -> Result<BlockResponse, HandleRequestError> {
 		let get_header = attributes.contains(BlockAttributes::HEADER);
 		let get_body = attributes.contains(BlockAttributes::BODY);
 		let get_justification = attributes.contains(BlockAttributes::JUSTIFICATION);
 
 		let mut blocks = Vec::new();
-		let mut block_id = from_block_id;
 
 		let mut total_size: usize = 0;
-		while let Some(header) = self.client.header(block_id).unwrap_or(None) {
+		loop {
+			let header = match self.client.header(block_id) {
+				Ok(Some(h)) => h,
+				Ok(None) => {
+					// If someone send us a request with an ascending direction and we don't have
+					// the block, we should decrease its reputation.
+					if direction == Direction::Ascending {
+						reputation_changes.push(rep::INVALID_REQUEST);
+					}
```

Comment:
I'm not really sure what you're trying to achieve. Is `Ascending` used as a hint for the fact that the remote wants to sync from us as opposed to do an ancestry search?

In any way, I'd suggest to not decrease the reputation if for example a peer asks for blocks 100 to 200 but we only have blocks 100 to 150.

```suggestion
					// If someone send us a request with an ascending direction and we don't have
					// the block, we should decrease its reputation.
					if blocks.is_empty() && direction == Direction::Ascending {
						reputation_changes.push(rep::INVALID_REQUEST);
					}
```

Reply:
If someone asks us for block 5 to 500, but we only announced that we have import 400, this is an invalid request. 

Why I only do this for `Ascending`, we could have announced some block X and the other one wants to have from X descending 5 blocks, but in the meantime we have pruned this one, because it was a fork.

Reply:
That's not an invalid request really, as it does not cost us anything to not send the missing 100 blocks. Basically it says "Send me as many blocks as you have starting from 5"

Reply:
Okay, I will remove this then.